### PR TITLE
Refactor test and benchmark infrastructure

### PR DESCRIPTION
**Pull request by ndryden**
This completely revamps the testing and benchmarking infrastructure in Aluminum to be more modular, cleaner, and comprehensive.

Currently a work-in-progress. Remaining todos:

- [x] Support point-to-point testing in `test_ops`.
- [x] Support specifying a particular algorithm (or all algorithms) to test.
- [x] Update or eliminate remaining tests (`test_exchange`, `test_multi_nballreduces`, `test_rma_halo_exchange`, `test_rma_ring`, `test_transfer_from_one`, `test_transfer_to_one`).
- [x] Fix `--dump-on-error` for in-place rooted operations (e.g., gather).
- [x] Port removed benchmarking infrastructure.
- [x] Unify operation benchmarks.
- [x] Update or eliminate remaining benchmarks.

### CODE REVIEW

**Comment by ndryden** [comment]
This is ready for review. Some notes:
- Building tests or benchmarks now takes quite a long time.
- Tests and benchmarks are no longer built by default, and must be explicitly requested with the `ALUMINUM_ENABLE_TESTS/BENCHMARKS` CMake flag. Debug builds enable tests.
- Nearly all tests have been unified into `test_ops`; likewise benchmarks and `benchmark_ops`. The command-line flags should be mostly self-explanatory. Some tests/benchmarks that I didn't think were especially useful have been eliminated. The remaining ones have been ported, but I haven't tried to update them significantly.
- Notably, the ping-pong benchmark is now basically done automatically in `benchmark_ops` when you request either a send or recv operator.
- When requesting `send` or `recv` operators, some ranks will be assigned to do sends, others receives, as otherwise things don't make sense.
- You can request the default algorithm, a specific algorithm, or all algorithms be tested with the `--algorithm` flag. You can only benchmark a single algorithm at a time (no `all` option).
- Commentary on code structure/design appreciated.

Some things for future PRs:
- I'd like to support benchmarking multiple algorithms in one invocation, but the `OpProfile` helper class will need to be reengineered to better track which algorithm is used.
- Everything related to algorithms is somewhat ugly because Aluminum makes things difficult -- I'd like to reengineer that part of Aluminum somewhat.
- A lot of the template metaprogramming stuff done for this seems like it actually belongs in Aluminum proper.
- I'd like to incorporate the wait kernel into `benchmark_ops` to hide launch overheads.

File: benchmark/benchmark_ops.cpp [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+      std::cout << options.help() << std::endl;
+    }
+    test_fini_aluminum();
+    std::exit(0);
+  }
+
+  // Simple validation.
+  if (!parsed_opts.count("op")) {
+    std::cerr << "Must provide an operator to test" << std::endl;
+    std::abort();
```
Comment (benson31) [inline benchmark/benchmark_ops.cpp:233]:
```suggestion
    test_fini_aluminum();
    return EXIT_FAILURE;
```

File: benchmark/benchmark_ops.cpp [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+  }
+  if (!parsed_opts.count("backend")) {
+    std::cerr << "Must provide a backend to use" << std::endl;
+    std::abort();
```
Comment (benson31) [inline benchmark/benchmark_ops.cpp:237]:
```suggestion
    test_fini_aluminum();
    return EXIT_FAILURE;
```

File: include/aluminum/utils.hpp [hunk 3]
```diff
@@ -44,8 +44,9 @@ inline double get_time() {
  *
  * This is mostly meant to help with vector collectives.
  */
-inline std::vector<size_t> excl_prefix_sum(const std::vector<size_t>& v) {
-  std::vector<size_t> r = std::vector<size_t>(v.size(), size_t{0});
+template <typename T>
+inline std::vector<T> excl_prefix_sum(const std::vector<T>& v) {
+  std::vector<T> r = std::vector<T>(v.size(), T{0});
```
Comment (benson31) [inline include/aluminum/utils.hpp:49]:
```suggestion
  auto r = std::vector<T>(v.size(), T{0});
```

File: test/test_exchange.cpp [hunk 4]
```diff
@@ -1,123 +1,149 @@
...
-  } else if (backend == "NCCL") {
-    test_correctness<Al::NCCLBackend>();
-#endif
-#ifdef AL_HAS_MPI_CUDA
-  } else if (backend == "MPI-CUDA") {
-    std::cerr << "Point-to-point not supported on MPI-CUDA backend." << std::endl;
+  // Simple validation.
+  if (!parsed_opts.count("backend")) {
+    std::cerr << "Must provide a backend to use" << std::endl;
     std::abort();
```
Comment (benson31) [inline test/test_exchange.cpp:144]:
```suggestion
    test_fini_aluminum();
    return EXIT_FAILURE;
```

File: test/test_ops.cpp [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  }
+
+  if (parsed_opts.count("hang-rank")) {
+    hang_for_debugging(parsed_opts["hang-rank"].as<int>());
+  }
+
+  // Simple validation.
+  if (!parsed_opts.count("op")) {
+    std::cerr << "Must provide an operator to test" << std::endl;
+    std::abort();
```
Comment (benson31) [inline test/test_ops.cpp:332]:
```suggestion
    test_fini_aluminum();
    return EXIT_FAILURE;
```

File: test/test_ops.cpp [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+  }
+  if (!parsed_opts.count("backend")) {
+    std::cerr << "Must provide a backend to use" << std::endl;
+    std::abort();
```
Comment (benson31) [inline test/test_ops.cpp:336]:
```suggestion
    test_fini_aluminum();
    return EXIT_FAILURE;
```

**Review by benson31 — APPROVED** [review]
LGTM. This should be good to merge; I'll test ROCm later on.

**Pull request merged**